tracer/src/Datadog.Trace/Debugger/Expressions/ProbeExpressionParser.cs

```diff
@@ -0,0 +1,483 @@
+            {
+                AddError(reader.Value?.ToString() ?? "N/A", e.Message);
+                return ReturnDefaultValueExpression();
+            }
+        }
+
+        return null;
+    }
+
+    private void AddError(string expression, string error)
+    {
+        (_errors ??= new List<EvaluationError>()).Add(new EvaluationError { Expression = expression, Message = error });
+    }
+
+    private bool ConditionalRead(JsonTextReader reader, bool shouldAdvanceReader)
+    {
+        return !shouldAdvanceReader || reader.Read();
+    }
+
+    private void SetReaderAtExpressionStart(JsonTextReader reader)
+    {
+        while (reader.Read())
+        {
+            if (reader.TokenType != JsonToken.PropertyName)
+            {
+                continue;
+            }
+
+            if ((reader.Value?.ToString() == "json"))
+            {
+                reader.Read();
+                reader.Read();
+                return;
+            }
+        }
+
+        throw new InvalidOperationException("Invalid json file");
+    }
+
+    private Expression HandleReturnType(Expression finalExpr)
+    {
+        if (typeof(T).IsAssignableFrom(finalExpr.Type))
+        {
+            return finalExpr;
+        }
+
+        if (typeof(T) != typeof(string))
+        {
+            // let the caller throw the correct assign exception
+            return finalExpr;
+        }
+
+        // for string, call ToString or return the type name
+        if (SupportedTypesService.IsSafeToCallToString(finalExpr.Type))
+        {
+            finalExpr = Expression.Call(finalExpr, GetMethodByReflection(typeof(object), nameof(object.ToString), Type.EmptyTypes));
+        }
+        else
+        {
+            finalExpr = Expression.Constant(finalExpr.Type.FullName, typeof(string));
+        }
+
+        return finalExpr;
+    }
+
+    private ExpressionBodyAndParameters ParseProbeExpression(string expressionJson, ScopeMember @this, ScopeMember[] argsOrLocals)
+    {
+        @this.Type ??= @this.Value?.GetType();
+        if (string.IsNullOrEmpty(expressionJson) || argsOrLocals == null || @this.Type == null)
+        {
+            throw new ArgumentException($"{nameof(ParseProbeExpression)} has been called with an invalid argument");
+        }
+
+        var scopeMembers = new List<ParameterExpression>();
+        var expressions = new List<Expression>();
+        var thisParameterExpression = Expression.Parameter(@this.GetType());
+        var thisVariable = Expression.Variable(@this.Type, "this");
+        expressions.Add(Expression.Assign(thisVariable, Expression.Convert(Expression.Field(thisParameterExpression, "Value"), @this.Type)));
+        scopeMembers.Add(thisVariable);
+
+        var argsOrLocalsParameterExpression = Expression.Parameter(argsOrLocals.GetType());
+
+        for (var index = 0; index < argsOrLocals.Length; index++)
+        {
+            if (argsOrLocals[index].Type == null)
+            {
+                break;
```

Comment:
Are you sure we want to `break` instead of `continue`?
And what will happen if we miss/skip certain arg/locals? how will it affect the flow?